### DISCUSSION TITLE
Private documentation with auth or jwt token check

### DISCUSSION THREAD

**Discussion by alexstock1** [withastro/starlight#1348] [Feature Requests]
### What version of `starlight` are you using?

4.0.0

### What is your idea?

I have a commercial product and would like to implement the capability to authenticate by JWT tokens. This would enable users to access the documentation only after a successful check with the token

### Why is this feature necessary?

Since the documentation is private and should not be publicly available I would like to add this check

### Participation

- [ ] I am willing to submit a pull request for this feature.

**Comment by dealmengor** [comment]
are you able to implement this?

**Comment by delucis** [comment]
Sharing a link to a new template for Starlight + authentication built by @TheOtterlord: https://github.com/TheOtterlord/starlight-auth

**Comment by adrianAzoitei** [comment]
Was anyone able to run starlight-auth with node? Seems to work as expected on Vercel, but I cannot get it to run on Node yet. Tried compiling server/indext.ts but couldn't figure it out yet.

**Reply by Fryuni** [reply]
`starlight-auth` uses a Vercel edge middleware for the auth on docs pages. To do it in Node, you'd need to build your Astro project to Node in middleware mode, write your own Node server doing the auth, and forward it to the Astro project if it passes.

**Comment by HiDeoo** [comment]
Sharing a new [example](https://github.com/HiDeoo/starlight-better-auth-example) I created that uses the [Better Auth](https://www.better-auth.com/) library to handle user authentication: users can create an account, sign in, and sign out — some pages are protected and require the user to be authenticated to access them or even be visible in the documentation sidebar.

**Comment by jemand771** [comment]
I'm making my life easy by using "proxy auth" - i.e. traffic hits my IdP which does the authentication and session management and almost acts like a reverse proxy but with extra steps. Finally, (if signed in) it passes on some user data via http headers to my astro site in ssr mode. Those headers are trivial to read from a middleware, **but**

the part I'm struggling with is how to get a markdown page's props from my middleware. Sure, checking if the slug starts with a certain string is _okay_ I guess but ideally I'd like to have a page prop in my frontmatter that controls the "confidentiality" level of that page - i.e. whether it requires signin, who can view it, and so on

**Reply by delucis** [reply]
In the middleware you could use Astro’s [`getEntry()`](https://docs.astro.build/en/reference/modules/astro-content/#getentry) utility I think to get the frontmatter for the current page. Unfortunately at the time middleware runs, Starlight’s own logic hasn’t kicked in yet, so there’s not a simple way to just pass the current page’s details automatically, but given the slug, you should be able to retrieve the entry I think.

**Reply by jemand771** [reply]
oh wow, I didn't know that would just work like that

I resorted to `getRouteBySlugParam(context.params.slug)` (from an incredibly dirty `import { getRouteBySlugParam } from '../node_modules/@astrojs/starlight/utils/routing';`) but using the officially supported way is of course way better

huge thanks!

**Reply by delucis** [reply]
Heh, I didn’t even think about that internal, but I won’t tell anyone you’re using it 😁

**Reply by Fryuni** [reply]
It's there for SSR pages Chris. Gotta find the pages dynamically 🕵🏼

**Comment by delucis** [comment]
As there are a couple of examples already shared here for how to do this, I’m going to closes this feature request as I don’t think there are specific changes to the Starlight internals needed to support this.
